test/Microsoft.AspNetCore.Razor.Test.Common/Language/Legacy/SyntaxTreeNodeWriter.cs

```diff
@@ -0,0 +1,231 @@
+            WriteSpan(span);
+            WriteSeparator();
+            Write(chunkGenerator.LookupText);
+            WriteSeparator();
+            Write(chunkGenerator.DirectiveText);
+            WriteSeparator();
+            Write(chunkGenerator.TypePattern);
+            WriteSeparator();
+            Write(chunkGenerator.AssemblyName);
+        }
+
+        public override void VisitTagHelperPrefixDirectiveSpan(TagHelperPrefixDirectiveChunkGenerator chunkGenerator, Span span)
+        {
+            WriteSpan(span);
+            WriteSeparator();
+            Write(chunkGenerator.Prefix);
+            WriteSeparator();
+            Write(chunkGenerator.DirectiveText);
+        }
+
+        public override void VisitStatementSpan(StatementChunkGenerator chunkGenerator, Span span)
+        {
+            WriteSpan(span);
+        }
+
+        public override void VisitDirectiveToken(DirectiveTokenChunkGenerator chunkGenerator, Span span)
+        {
+            WriteSpan(span);
+            WriteSeparator();
+            Write(chunkGenerator.Descriptor.Kind);
+            WriteSeparator();
+            Write(chunkGenerator.Descriptor.Name);
+            WriteSeparator();
+            Write($"Optional: {chunkGenerator.Descriptor.Optional}");
+        }
+
+        protected void WriteBlock(Block block)
+        {
+            WriteIndent();
+            Write($"{block.Type} block");
+            WriteSeparator();
+            Write(block.ChunkGenerator.GetType().Name);
+            WriteSeparator();
+            Write(block.Length);
+            WriteSeparator();
+            WriteSourceLocation(block.Start);
+        }
+
+        protected void WriteSpan(Span span)
+        {
+            WriteIndent();
+            Write($"{span.Kind} span");
+            WriteSeparator();
+            Write(span.ChunkGenerator.GetType().Name);
+            WriteSeparator();
+            Write(span.EditHandler);
+            WriteSeparator();
+            Write(span.Content);
+            WriteSeparator();
+            WriteSourceLocation(span.Start);
+        }
+
+        protected void WriteSourceLocation(SourceLocation location)
+        {
+            Write(location);
+        }
+
+        protected void WriteLocationTaggedString(LocationTagged<string> item)
+        {
+            Write(item.ToString("F", null));
+        }
+
+        protected void WriteIndent()
+        {
+            for (var i = 0; i < Depth; i++)
+            {
+                for (var j = 0; j < 4; j++)
+                {
+                    Write(' ');
+                }
+            }
+        }
+
+        protected void WriteSeparator()
+        {
+            Write(" - ");
+        }
+
+        protected void WriteNewLine()
+        {
+            _writer.WriteLine();
+        }
+
+        protected void Write(object value)
+        {
+            if (value is string stringValue)
+            {
+                stringValue = stringValue.Replace(Environment.NewLine, "LF");
```

Comment:
Representing new lines within span content etc as LF so we don't compromise on proper indentation aka readability.